server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateCache.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway.remote;
+
+import org.opensearch.cluster.ClusterName;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.coordination.ClusterStateTermVersion;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Cache to Remote Cluster State based on term-version check. The current implementation
+ * caches the last highest version of cluster-state that was downloaded from cache.
+ *
+ * @opensearch.internal
+ */
+public class RemoteClusterStateCache {
+
+    private final AtomicReference<ClusterState> clusterStateFromCache = new AtomicReference<>();
+
+    public ClusterState getState(String clusterName, ClusterMetadataManifest manifest) {
+        ClusterStateTermVersion manifestStateTermVersion = new ClusterStateTermVersion(
+            new ClusterName(clusterName),
+            manifest.getClusterUUID(),
+            manifest.getClusterTerm(),
+            manifest.getStateVersion()
+        );
+
+        ClusterState cache = clusterStateFromCache.get();
+        if (cache != null) {
+            ClusterStateTermVersion cacheStateTermVersion = new ClusterStateTermVersion(
+                new ClusterName(clusterName),
+                cache.metadata().clusterUUID(),
+                cache.term(),
+                cache.version()
+            );
```

Comment:
instead of creating this new object, it can be tuple in the atomic reference for ClusterStateTermVersion & ClusterState

Reply:
Changed it to cache as tuple of  ClusterStateTermVersion & ClusterState